Ensembl/ensembl-utils
# Rework to allow use of metadata

Maybe too many things changed, but I needed those to ensure I can use a metadata with :
- `DBConnection` can now be created without a schema (`dump_dir` no longer mandatory)
- `DBConnection` accepts an `Metadata` object instead of relying on an SQL file and a reflection
- Add 2 methods to `DBConnection`: `create_all_tables` and `create_table` (requires the metadata to be set as above)
- If both a `dump_dir` and a `metadata` are provided to a `UnitTestDB`, the schema is made from the `metadata` and the data is loaded from the `dump_dir` 
- Add a context manager to properly dispose of the `UnitTestDB` (`UnitTestDBContext`)

Tests:
- Add a simple mock base and mock_metadata
- When possible, use a context `session_scope` or `test_session_scope` to avoid hanging connections (unless we are specifically testing for that)
- Both SQLalchemy and MySQL should work
- Add an argument `tmp_path` to the `UnitTestDB` when asked to create an SQLite database test file

Usage:
Say you want to test interactions with a whole schema described in a given metadata. You can do so with:
```
with UnitTestDBContext(server_url, metadata=metadata, tmp_path=tmp_path, name="test_metadata") as test_db:
    with test.db.dbc.test_session_scope() as session:
        results = session.execute(text("SELECT * FROM foo"))
```
The first context ensures the database is properly disposed of no matter what, and the second ensures the changes are rolledback and the connection returned.
